cranelift/wasm/src/state.rs

```diff
@@ -241,6 +242,52 @@ pub struct FuncTranslationState {
     // `FuncEnvironment::make_direct_func()`.
     // Stores both the function reference and the number of WebAssembly arguments
     functions: HashMap<FuncIndex, (ir::FuncRef, usize)>,
+
+    // TODO: maybe split into two separate members, latest_tagged_index and latest_tag. but having together makes sense due to the Option, whether each member is Some or None should always be the same
+
+    // In each (C) function, stack arrays are allocated contiguously. We want
+    // to deterministically guarantee that no two consecutive stack allocations
+    // share the same MTE tag. Therefore, we generate an initial MTE tag
+    // randomly (at runtime, using MTE's IRG), and deterministically generate
+    // the next tags using that initial tag. We store the lastest random tag,
+    // which we always need to calculate the next random tag.
+    latest_tagged_index: Option<Value>,
+}
+
+const MTE_NON_TAG_BITS_MASK: i64 = 0xF0FF_FFFF_FFFF_FFFFu64 as i64;
+const MTE_TAG_BITS_MASK: i64 = 0x0F00_0000_0000_0000;
+
+/// Take the 4-bit MTE tag from `from_index` and insert it into `to_index`.
+fn insert_tag_from_index_into_index(
+    from_index: Value,
+    to_index: Value,
+    builder: &mut FunctionBuilder,
+) -> Value {
+    // tag = from & 0x0F00... (keep only tag)
+    let tag = builder.ins().band_imm(from_index, MTE_TAG_BITS_MASK);
+
+    // to = to & 0xF0FF... (remove tag, keep rest)
+    let to_index = builder.ins().band_imm(to_index, MTE_NON_TAG_BITS_MASK);
+
+    // to | tag
+    builder.ins().bor(to_index, tag)
+}
+
+impl FuncTranslationState {
+    /// Tag the `index` with an MTE tag, and return the tagged index.
+    /// Contiguous stack allocations are guaranteed to have different random
+    /// tags.
+    pub fn tag_index(&mut self, index: Value, builder: &mut FunctionBuilder) -> Value {
+        let new_tagged_index = match self.latest_tagged_index.take() {
+            None => builder.ins().arm64_irg(index),
+            Some(previous_tagged_index) => {
+                let index = insert_tag_from_index_into_index(previous_tagged_index, index, builder);
+                builder.ins().arm64_add_to_tag(index, 1)
+            }
+        };
+        self.latest_tagged_index = Some(new_tagged_index);
+        new_tagged_index
+    }
```

Comment:
@martin-fink here's the core of the implementation. Do you see any more optimized ways e.g. for implementing `insert_tag_from_index_to_index`, maybe with less instructions?